python/ray/serve/deployment_executor_node.py

```diff
@@ -0,0 +1,78 @@
+from typing import Any, Dict, List
+
+from ray.experimental.dag import DAGNode
+from ray.serve.deployment_method_executor_node import DeploymentMethodExecutorNode
+from ray.experimental.dag.constants import DAGNODE_TYPE_KEY, PARENT_CLASS_NODE_KEY
+from ray.experimental.dag.format_utils import get_dag_node_str
+
+
+class DeploymentExecutorNode(DAGNode):
+    """The lightweight executor DAGNode of DeploymentNode that optimizes for
+    efficiency.
+
+        - We need Ray DAGNode's traversal and replacement mechanism to deal
+            with deeply nested nodes as args in the DAG
+        - Meanwhile, __init__, _copy_impl and _execute_impl are on the critical
+            pass of execution for every request.
+
+    Therefore for serve we introduce a minimal weight node as the final product
+    of DAG transformation, and will be used in actual execution as well as
+    deployment.
+    """
+
+    def __init__(
+        self,
+        deployment_handle,
+        dag_args,  # Not deployment init args
+        dag_kwargs,  # Not deployment init kwargs
+    ):
+        self._deployment_handle = deployment_handle
+        super().__init__(dag_args, dag_kwargs, {}, {})
+
+    def _copy_impl(
+        self,
+        new_args: List[Any],
+        new_kwargs: Dict[str, Any],
+        new_options: Dict[str, Any],
+        new_other_args_to_resolve: Dict[str, Any],
+    ):
+        return DeploymentExecutorNode(
+            self._deployment_handle,
+            new_args,
+            new_kwargs,
+        )
+
+    def _execute_impl(self, *args, **kwargs):
+        """Does not call into anything or produce a new value, as the time
+        this function gets called, all child nodes are already resolved to
+        ObjectRefs.
+        """
+        return self._deployment_handle
+
+    def __getattr__(self, method_name: str):
+        return DeploymentMethodExecutorNode(
+            method_name,
+            (),
+            {},
+            other_args_to_resolve={
+                PARENT_CLASS_NODE_KEY: self,
+            },
+        )
```

Comment:
is this valid? I don't think we should ever be calling `getattr` on this at execution time, right? It should already be replaced with a deployment method node.

Reply:
This method is actually only called at our own transformation time because we need to interpret `deployment.method` nodes, at execution time the only method matters is `_execute_impl` and `_copy_impl`

Reply:
I do think this PR reveals a bigger issue. We don't have an architecture/design doc for how Ray DAG works, what's all about `_execute_impl`, `_copy_impl`, `_other_args_to_resolve`... this is making people contribute and review relevant doc.